src/core/meshcop/joiner_router.hpp

```diff
@@ -26,10 +26,230 @@
+     *
+     * @returns The pointer to the parent otInstance structure.
+     *
+     */
+    otInstance *GetInstance(void);
+
+    /**
+     * This method returns the Joiner UDP Port.
+     *
+     * @returns The Joiner UDP Port number .
+     *
+     */
+    uint16_t GetJoinerUdpPort(void);
+
+    /**
+     * This method sets the Joiner UDP Port.
+     *
+     * @param[in]  The Joiner UDP Port number.
+     *
+     * @retval kThreadError_None    Successfully set the Joiner UDP Port.
+     *
+     */
+    ThreadError SetJoinerUdpPort(uint16_t aJoinerUdpPort);
+
+private:
+    enum
+    {
+        kDelayJoinEnt = 50,  ///< milliseconds
+    };
+
+    static void HandleNetifStateChanged(uint32_t aFlags, void *aContext);
+    void HandleNetifStateChanged(uint32_t aFlags);
+
+    static void HandleUdpReceive(void *aContext, otMessage *aMessage, const otMessageInfo *aMessageInfo);
+    void HandleUdpReceive(Message &aMessage, const Ip6::MessageInfo &aMessageInfo);
+
+    static void HandleRelayTransmit(void *aContext, otCoapHeader *aHeader, otMessage *aMessage,
+                                    const otMessageInfo *aMessageInfo);
+    void HandleRelayTransmit(Coap::Header &aHeader, Message &aMessage, const Ip6::MessageInfo &aMessageInfo);
+
+    static void HandleJoinerEntrustResponse(void *aContext, otCoapHeader *aHeader, otMessage *aMessage,
+                                            const otMessageInfo *aMessageInfo, ThreadError result);
+    void HandleJoinerEntrustResponse(Coap::Header *aHeader, Message *aMessage,
+                                     const Ip6::MessageInfo *aMessageInfo, ThreadError result);
+
+    static void HandleTimer(void *aContext);
+    void HandleTimer(void);
+
+    ThreadError DelaySendingJoinerEntrust(const Ip6::MessageInfo &aMessageInfo, const JoinerRouterKekTlv &aKek);
+    void SendDelayedJoinerEntrust(void);
+    ThreadError SendJoinerEntrust(Message &aMessage, const Ip6::MessageInfo &aMessageInfo);
+
+    ThreadError GetBorderAgentRloc(uint16_t &aRloc);
+
+    Ip6::NetifCallback mNetifCallback;
+
+    Ip6::UdpSocket mSocket;
+    Coap::Resource mRelayTransmit;
+    ThreadNetif &mNetif;
+
+    Timer mTimer;
+    MessageQueue mDelayedJoinEnts;
+
+    uint16_t mJoinerUdpPort;
+
+    bool mIsJoinerPortConfigured : 1;
+    bool mExpectJoinEntRsp : 1;
+};
+
+/**
+ * This class implements functionality required for delaying JOIN_ENT.ntf messages.
+ *
+ */
+class DelayedJoinEntHeader
+{
+public:
+    /**
+     * Default constructor for the object.
+     *
+     */
+    DelayedJoinEntHeader(void) { memset(this, 0, sizeof(*this)); }
+
+    /**
+     * This constructor initializes the object with specific values.
+     *
+     * @param[in]  aSendTime     Time when the message shall be sent.
+     * @param[in]  aDestination  IPv6 address of the message destination.
+     *
+     */
+    DelayedJoinEntHeader(uint32_t aSendTime, Ip6::MessageInfo &aMessageInfo, const uint8_t *aKek) {
```

Comment:
we should address this kind of format updates together in a separate PR (if we decide to follow that rule), do you agree?